kvserver: fix race on Replica.tenantLimiter

pav-kv opened the pull request:
When Replica is being destroyed, it can set tenantLimiter to nil. However, all accessors assume synchronization through IsInitialized atomic check and that this field can't be modified after the replica is initialized.

There is no particular need in nullifying this limiter, the replica is going away soon, so this commit removes this line in order to fix the data race.

Fixes #95345
Epic: none
Release note: none

pav-kv committed:
kvserver: fix race on Replica.tenantLimiter

When Replica is being destroyed, it can set tenantLimiter to nil. However, all
accessors assume synchronization through IsInitialized atomic check and that
this field can't be modified after the replica is initialized.

There is no particular need in nullifying this limiter, the replica is going
away soon, so this commit removes this line in order to fix the data race.

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95524)
<!-- Reviewable:end -->

pavelkalinnikov commented on pkg/kv/kvserver/replica_destroy.go:
```diff
@@ -132,7 +132,6 @@ func (r *Replica) postDestroyRaftMuLocked(ctx context.Context, ms enginepb.MVCCS
 	// Unhook the tenant rate limiter if we have one.
 	if r.tenantLimiter != nil {
 		r.store.tenantRateLimiters.Release(r.tenantLimiter)
-		r.tenantLimiter = nil
```
Alternatively, we could protect `tenantLimiter` with `Replica.mu`, but I don't see much benefit. WDYT?

tbg replied:
Both options seem a little iffy. I don't want us to introduce another mutex acquisition in the critical path but at the same time it's not trivial to rearrange the users of this into an existing critical section.

Use-after-Release on the `tenantLimiter` might work - I looked at the code a bit and it seems like it doesn't destroy or pool any memory; it really just unlinks the limiter from the parent.

Let's go with the second option (the one you have chosen) but add a test in `tenantrate` that proves that you can use a limiter after it's been released and that this will not crash and also no longer influence the surrounding aggregating metric.

pavelkalinnikov replied:
I wrote a test. It does not panic after `Release`, but the parent metric is still updated. The `Destroy` method removes the child metric from the parent's tree, but still leaves the child->parent reference and keeps updating the parent metric. This API doesn't quite make sense to me, but it is what it is.

I think it's fine for counter metrics if some in-flight `Inc`rements slip through, but with the `Gauge` metrics we can't reliably reset it to 0 (and update the parent so that it no longer accounts for this value), a racing change may result in a permanent non-zero leak to the aggregate.

pavelkalinnikov replied:
I don't know, but the whole fact that `Send` can race with `Replica` destruction, and operate on a half-destroyed one, seems iffy.

pavelkalinnikov replied:
I think that the `Replica` "state machine" is confusing. Many things assume that there are only 2 states: uninitialized and initialized (hence the `IsInitialized` checks in many places, including `Send`, and assumption that many things synchronize through it), but in reality there is also `destroyStatus` which breaks this dichotomy.

tbg replied:
@RaduBerinde for this usage here

https://github.com/cockroachdb/cockroach/blob/e6e092fefc2a5ba0ff6342728f7b14f51e789899/pkg/ccl/multitenantccl/tenantcostserver/metrics.go#L170-L193

isn't it weird that these are gauges in the first place? And also, I don't see this getting released anywhere (we hold on to a map entry for a tenant forever), I assume this is intentional. So this should not be affected by not resetting gauges to zero but please let me know if we're missing something.

tbg reviewed (changes requested):
LGTM mod adding a unit test as per comment.

pav-kv committed:
tenantrate,aggmetric: properly handle releasing gauges

Unfortunately we have to accept that child metrics continue to see usage
(at least for a short period of time) despite already having been
"Destroyed". The code assumed that a call to `Destroy()` would prevent
all future use of the metric, but this is not the case. To paper over
this, `Destroy` (for a gauge) was resetting to zero, but this would mean
that some callers who were careful to always decrement the gauge
themselves (for example, the quota pool's "pending waiters" metric)
would actually end up with a corrupted value of the gauge.

In a follow up commit we will stop pretending that `Destroy` destroys
the metric. Instead, we will rename to "Unlink" to make clear(er) that
this really only unlinks the child from the parent. It would be nice to
also make clear in the name that the other direction doesn't hold
(updates to child continue to update the parent) but I can't find a
better way to do this than to spell it out explicitly in the comment.

The semantics here aren't that fortunate, but it's a bigger re-work
to make this work better.

Epic: none
Release note: None

tbg committed:
aggmetric: Destroy -> Unlink

Epic: none
Release note: None

pavelkalinnikov commented on pkg/kv/kvserver/tenantrate/factory.go:
```diff
@@ -121,6 +121,7 @@ func (rl *LimiterFactory) Release(lim Limiter) {
 	}
 	if rcLim.refCount--; rcLim.refCount == 0 {
 		l.metrics.destroy()
+		l.qp.Close("released")
```
Nice

pavelkalinnikov commented:
LGTM

tbg commented:
I audited the callers to `Unlink()` and they're specific to the use of aggmetric in tenant storage metrics and I think there's only one more that needed to change (see new commit).
Unfortunately this didn't break any tests so wondering if you could backfill a unit test and also double check my analysis that nobody else even called `Unlink` on `Gauge{,Float64}` other than the two places we have now dealt with (tenant rate limiter and tenant store metrics).

pav-kv committed:
kvserver: add TenantsStorageMetrics release test

This test verifies that TenantsStorageMetrics is updated correctly when tenant
IDs are released.

Release note: none
Epic: none

pavelkalinnikov commented:
@tbg Added a test (also verified that it would fail with your penultimate commit). Yes, `Unlink` seems to be only used by the two places that we've covered.

pavelkalinnikov commented on pkg/kv/kvserver/tenantrate/factory.go:
```diff
@@ -121,6 +121,7 @@ func (rl *LimiterFactory) Release(lim Limiter) {
 	}
 	if rcLim.refCount--; rcLim.refCount == 0 {
 		l.metrics.destroy()
+		l.qp.Close("released")
```
Does this strictly prevent the race? The other end may notice the closed channel too late, and still write to the metric, no?

I wonder if the proper solution to this would be ref-counting not only at the `Replica` level, but at the request level. I.e. make it so that `Send` gets a ref-incremented `Limiter`, and `Releases` it in the end; instead of relying on `Replica` to do so.

pavelkalinnikov replied:
Ah, I see, we rely on the fact that there is only one `Gauge` metric which also happens to be related to synchronization. We close the channel, and then some time after the waiters decrement the counter. In the meantime, we only `Unlink`. The counter eventually will stabilize.

Any generic gauge addition would break things again. Worth noting this and TODOing.

tbg replied:
👍🏽 Did you want to add a TODO?

pavelkalinnikov replied:
I think you covered it well in the comment next to `r.tenantLimiter`. Injected a TODO there.

tbg commented on pkg/kv/kvserver/metrics.go:
```diff
@@ -2152,24 +2152,31 @@ func (sm *TenantsStorageMetrics) releaseTenant(ctx context.Context, ref *tenantM
+		&m.AbortSpanBytes,
+	} {
+		// Reset before unlinking, see Unlink.
+		(*gptr).Update(0)
+		(*gptr).Unlink()
+		*gptr = nil
```
I'm not opposed, your PR after all! For me personally this is ok as is.

pavelkalinnikov replied:
Leaving as is.

pavelkalinnikov commented:
TFTR!

pavelkalinnikov commented:
bors r=tbg

pavelkalinnikov commented:
bors retry
